maintainer/deploy_docs_via_travis.sh

```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+# Deploying docs from travis-ci.
+# See https://github.com/MDAnalysis/mdanalysis/issues/386
+# Script based on https://github.com/steveklabnik/automatically_update_github_pages_with_travis_example
+
+# Run this script from the top-level of the checked out git
+# repository. A github OAuth token must be available in the evironment
+# variable GH_TOKEN and is set up through the .travis.yml
+# env:global:secure parameter (encrypted with travis-ci's public key)/
+#
+# Additional environment variables set in .travis.yml
+#  GH_REPOSITORY     repo to full from and push to
+#  GH_DOC_BRANCH     branch from which the docs are built
+#  GIT_CI_USER       name of the user to push docs as
+#  GIT_CI_EMAIL      email of the user to push docs as
+#  MDA_DOCDIR        path to the docdir from top of repo
+#  MAINTAIN_DIR      path to maintainer/
+#  VERSION           version of MDAnalysis
+#
+# NOTE: If any of these environment variables are not set or 
+#       empty then the script will exit with and error (-o nounset).
+
+set -o errexit -o nounset
+
+function die () {
+    local msg="$1" err=${2:-1}
+    echo "ERROR: $msg [$err]"
+    exit $err
+}
+
+rev=$(git rev-parse --short HEAD)
+
+# the following tests should be superfluous because of -o nounset
+test -n "${GH_TOKEN}" || die "GH_TOKEN is empty: need OAuth GitHub token to continue" 100
+test -n "${GH_REPOSITORY}" || die "GH_REPOSITORY must be set in .travis.yml" 100
+test -n "${MDA_DOCDIR}" || die "MDA_DOCDIR must be set in .travis.yml" 100
+test -n "${MAINTAIN_DIR}" || die "MAINTAIN_DIR must be set in .travis.yml" 100
+test -n "${VERSION}" || die "VERSION must be set in .travis.yml" 100
+
+
+cd ${MDA_DOCDIR} || die "Failed to 'cd ${MDA_DOCDIR}'. Run from the top level of the repository"
+
+# move into $version subdirectory
+mkdir ../${VERSION} && mv * ../${VERSION}
+
+git init
+git config user.name "${GIT_CI_USER}"
+git config user.email "${GIT_CI_EMAIL}"
+
+mv ../${VERSION} $VERSION
+
+git remote add upstream "https://${GH_TOKEN}@${GH_REPOSITORY}"
+git fetch --depth 50 upstream ${GH_DOC_BRANCH} gh-pages
+git reset upstream/gh-pages
+
+# for dev, latest, home redirects
+mkdir dev latest
+python ${MAINTAIN_DIR}/update_json_stubs_sitemap.py
+touch .
+touch .nojekyll
+
+git add -A ${VERSION}/
+git add .nojekyll versions.json
+git add dev latest
+git add *.xml
+
+echo "ls *"
+ls *
+
+echo $(ls *.html | wc -l)
+
+if [ "$(ls *.html | wc -l)" -ge "1" ]; then
+    cd $VERSION && mdfiles=$(ls -d *) && cd -
+    for item in $mdfiles ; do
+        if [[ -f $item ]] || [[ -d $item ]]; then echo $item && git add $item; fi
+    done
+fi
```

Comment:
This looks a bit fragile and I am not quite sure why it's needed after `git add -A  $VERSION`.

Reply:
This adds the redirect files in the top level directory (`$mdfiles` is a list of all the files and directories inside `$VERSION`, and then we `cd -` to add the redirects). If we ``git add .`` it removes other `$VERSION` directories from previous commits, as it registers their absence. We could ignore removals but that doesn't help when we actually want to remove files. We could pull from the remote but I wasn't sure if this would result in merge conflicts. So this clumsy workaround.

I had to check if each `$item` exists because some directories don't contain `*.html` files, and apparently trying to `git add` a nonexistent item causes a fatal error and makes script exit.